### DISCUSSION TITLE
Statistics component redesign for direct recorder history usage

### DISCUSSION THREAD

**Discussion by ThomDietrich** [home-assistant/architecture#784] [General]
Hey all,

in this ticket I want to describe and discuss a significant change to the statistics component.

**TL;DR;** I intent to remove the internal (arguably redundant) state storage from the statistics component and rely on the recorder for state history. The change will resolve design restrictions and performance concerns. Additionally, it enables new features off the back of the component, e.g. statistics based automation triggers and conditions. This plan stands and falls with current limits and future plans for the recorder, hence this discussion thread.

Blocks https://github.com/home-assistant/core/issues/67627 https://github.com/home-assistant/core/pull/69700 https://github.com/home-assistant/core/pull/62624 and further contribution plans by me.

## Current state

The statistics component implements an internal memory (`deque`) to store state and time information. Every new state change is captured and considered for statistical computation. IF the recorder is active, historic data is used initially to pre-populate the sensor.

https://github.com/home-assistant/core/blob/df67a8cd4f8df91a153778009a74be1e3876ca53/homeassistant/components/statistics/sensor.py#L300-L301

### Problem/Limitation

The statistics component duplicates data storage and management functionality of the recorder. I recon this was originally motivated by the fact that the recorder was not enabled by default. Nowadays, the situation is quite different however, and I assume this restriction can be lifted. **Please challenge if you believe this assumption is wrong.**

I did not run comparisons but I expect the deque to be less performant than the recorder.

The current implementation limited me in the implementation of past PRs and limits future features (see below).

## Proposal

I want to work on a PR which

1. Removes local memory from the statistics component
2. Leverages the recorder history for statistical computation
3. Renders the pre-population logic in the statistics component useless.

The solution is successfully implemented in this custom-component: https://github.com/Limych/ha-average/blob/dev/custom_components/average/sensor.py

### Benefits

The change will ease the path for the following and more features:

- Statistical computation over longer time spans / sample counts
- Consideration of youngest state change outside the configured moving window (please ping if you are interested in contextual details)
- Access to historic information aside value and datetime
- Direct use of statistics as trigger or condition, in addition to entity definition

## Questions

Which future changes to the recorder are to be expected and which current limits might be blockers?

- Is it considered to not record certain entities?
- Is it considered to limit the persistence period for entities?
- Is it considered to reduce data granularity for entities?
- Are maximum periods or data reductions in place today?
- Did I miss a critical limitation of the recorder / a benefit of the current solution?
- In the grand schema of things, might it be necessary and/or meaningful to migrate some of the elements from the statistics component closer to the recorder?
- How might this relate to "long-term statistics" and future development plans?

Looking forward to your input. Thanks!

CC @bdraco @fabaff @emontnemery @MartinHjelmare

**Comment by ThomDietrich** [comment]
@bdraco was hoping for your input specifically. Cheers

**Comment by emontnemery** [comment]
In general this seems like an interesting idea. Some thoughts on your questions:
* Is it considered to not record certain entities?
  * Users can exclude entities from recording, there's also some thought around excluding sensors from being recorded, related to long term statistics
* Is it considered to limit the persistence period for entities?
  * Sure, and this is user configurable
* Is it considered to reduce data granularity for entities?
  * There is the `significant_change` feature. We currently record all state changes, regardless of if they're significant or not.
* Are maximum periods or data reductions in place today?
  * Unless configured otherwise, the recorder retains entity states for 10 days.
* Did I miss a critical limitation of the recorder / a benefit of the current solution?
  * It may be challenging to implement a performant implementation if a database search is done each time a statistics sensor needs to be updated, whereas the `deque` class is very fast.
* In the grand schema of things, might it be necessary and/or meaningful to migrate some of the elements from the statistics component closer to the recorder?
  * Hard to say at this point
* How might this relate to "long-term statistics" and future development plans?
  * One goal, which is achievable thanks to long-term statistics, is to have a shorter retain period for sensors or not record them at all. This would for sure impact the `statistics` integration.

If we go down this path, I think the `statistics` component could call a variant of `recorded.is_entity_recorded` to find out if an entity is being recorded and for how long data is retained. If the entity is not recorded, or if the retainment period is to short `statistics` can then log a warning + open a repairs issue.
